src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -0,0 +1,216 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ingestcarbon
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/m3db/m3/src/cmd/services/m3coordinator/ingest"
+	"github.com/m3db/m3/src/metrics/carbon"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3x/instrument"
+	m3xserver "github.com/m3db/m3x/server"
+	xsync "github.com/m3db/m3x/sync"
+	xtime "github.com/m3db/m3x/time"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	// Used for parsing carbon names into tags.
+	carbonSeparatorByte  = byte('.')
+	carbonSeparatorBytes = []byte{carbonSeparatorByte}
+
+	// Number of pre-formatted key names to generate in the init() function.
+	numPreFormattedKeyNames = 100
+	// Should never be modified after init().
+	preFormattedKeyNames = [][]byte{}
+
+	errCannotGenerateTagsFromEmptyName = errors.New("cannot generate tags from empty name")
+)
+
+// Options configures the ingester.
+type Options struct {
+	InstrumentOptions instrument.Options
+	WorkerPool        xsync.PooledWorkerPool
+}
+
+// NewIngester returns an ingester for carbon metrics.
+func NewIngester(
+	downsamplerAndWriter ingest.DownsamplerAndWriter,
+	opts Options,
+) (m3xserver.Handler, error) {
+	// TODO(rartoul): Validate options.
+	return &ingester{
+		downsamplerAndWriter: downsamplerAndWriter,
+		opts:                 opts,
+	}, nil
+}
+
+type ingester struct {
+	downsamplerAndWriter ingest.DownsamplerAndWriter
+	opts                 Options
+}
+
+// TODO(rartoul): Emit metrics
+func (i *ingester) Handle(conn net.Conn) {
+	var (
+		wg     = sync.WaitGroup{}
+		s      = carbon.NewScanner(conn, i.opts.InstrumentOptions)
+		logger = i.opts.InstrumentOptions.Logger()
+	)
+
+	logger.Debug("handling new carbon ingestion connection")
+	for s.Scan() {
+		name, timestamp, value := s.Metric()
+		// Copy name since scanner bytes are recycled.
+		// TODO(rartoul): Pool.
+		name = append([]byte(nil), name...)
+
+		wg.Add(1)
+		i.opts.WorkerPool.Go(func() {
+			datapoints := []ts.Datapoint{{Timestamp: timestamp, Value: value}}
+			// TODO(rartoul): Pool.
+			tags, err := GenerateTagsFromName(name)
+			if err != nil {
+				logger.Errorf("err generating tags from carbon name: %s, err: %s",
+					string(name), err)
+				return
+			}
+
+			// TODO(rartoul): Context with timeout?
+			err = i.downsamplerAndWriter.Write(
+				context.Background(), tags, datapoints, xtime.Second)
+			if err != nil {
+				logger.Errorf("err writing carbon metric: %s, err: %s",
+					string(name), err)
+			}
+
+			wg.Done()
```

Comment:
This `wg.Done()` is not called if early return up top.

Perhaps just pay the cost of using `defer wg.Done()` at the top of the goroutine?

Reply:
good catch.